### ISSUE TITLE
تمرین سوم

### ISSUE THREAD

**Issue by romosij200218-prog** [Amir13849271741537/Machine-learning-#31]
Tania:
شاخص‌های ارزیابی مدل Naive Bayes
• Accuracy (دقت): 0.977
• Precision (دقت مثبت): 0.979
• Recall (بازآوری): 0.977
• F1-score: 0.977

ماتریس درهم‌ریختگی (Confusion Matrix)

[[19,  0,  0],
[ 0, 12,  1],
[ 0,  0, 13]]
این نتایج نشان می‌دهد که مدل Naive Bayes عملکرد بسیار خوبی روی دیتاست Iris داشته است
استفاده از الگوریتم Naive Bayes

الگوریتم Naive Bayes یک روش طبقه‌بندی ساده و قدرتمند است که بر اساس قانون احتمال بیز کار می‌کند.
این الگوریتم فرض می‌کند ویژگی‌ها مستقل از هم هستند؛ به همین دلیل به آن Naive (ساده‌لوحانه) می‌گویند.

برای این تمرین از نسخه Gaussian Naive Bayes استفاده کردیم؛ این نسخه مخصوص داده‌هایی است که ویژگی‌های عددی با توزیع نرمال دارند.

۲. دیتاست مورد استفاده: Iris

دیتاست Iris شامل 150 نمونه گل است با 4 ویژگی:
• طول کاسبرگ (sepal length)
• عرض کاسبرگ
• طول گلبرگ (petal length)
• عرض گلبرگ
• برچسب کلاس شامل 3 نوع گل:
• Iris-setosa
• Iris-versicolor
• Iris-virginic
۳. مراحل انجام کار

مرحله ۱: جدا کردن داده‌ها به دو بخش train و test

70٪ داده‌ها برای آموزش
30٪ داده‌ها برای تست مدل

مرحله ۲: آموزش مدل GaussianNB

مدل روی داده‌های آموزشی fit شد.

مرحله ۳: پیش‌بینی

مدل روی داده‌های تست پیش‌بینی کرد.

مرحله ۴: ارزیابی با شاخص‌ها
۴. توضیح کامل شاخص‌های ارزیابی
1. Accuracy (دقت کلی)
نشان می‌دهد چه مقدار از نمونه‌های تست را درست پیش‌بینی کرده است

Accuracy = 0.977
یعنی 97.7٪ پیش‌بینی‌ها درست بوده‌اند
Precision (دقت مثبت
می‌گوید وقتی مدل یک کلاس را پیش‌بینی کرده، چقدر درست بوده است
3. Recall (بازآوری)
نشان می‌دهد از تمام نمونه‌های واقعی یک کلاس، چند درصد را درست پیدا کرده است

Recall = 0.977
4. F1-score
میانگینی از precision و recall است و وقتی این دو نامتوازن باشند، اهمیت دارد

F1 = 0.977
۵. ماتریس درهم‌ریختگی (Confusion Matrix

[[19, 0, 0],  
[ 0,12, 1],  
[ 0, 0,13]]
تفسیر:
• کلاس ۱ (Setosa): همه 19 نمونه درست تشخیص داده شده‌اند.
• کلاس ۲ (Versicolor): 12 درست، 1 اشتباه → به کلاس سوم پیش‌بینی شده
• کلاس ۳ (Virginica): همه 13 درست تشخیص داده شده‌اند.

یعنی تنها یک خطا در کل پیش‌بینی‌ها رخ داده!
۶. نتیجه‌گیری نهایی

مدل Naive Bayes برای دیتاستی مثل Iris که ویژگی‌ها نسبتاً نرمال و واضح هستند، بسیار عملکرد خوبی دارد و دقت بالای 97٪ به‌دست آورده است.

تمرین سوم ،موضوع ۳: پیش‌بینی ریسک پزشکی با Linear Regression

برای این مسئله از دیتاست معروف Diabetes موجود در scikit-learn استفاده می‌کنیم.
این دیتاست دقیقاً مناسب “پیش‌بینی ریسک پزشکی” است، زیرا:
• داده‌های واقعی پزشکی دارد
• هدف مدل: پیش‌بینی شدت پیشرفت بیماری دیابت
• خروجی یک مقدار عددی است → مناسب برای Linear Regression

۱. معرفی دیتاست Diabetes

این دیتاست شامل ۱۰ ویژگی پزشکی است مانند:
• سن
• BMI
• فشار خون
• قند خون
• چندین شاخص بیوشیمی

و خروجی:
▶ عدد ریسک/پیشرفت بیماری دیابت طی زمان

۲. مراحل انجام با Linear Regression
در اجرای واقعی (که به خاطر محدودیت لحظه‌ای ابزار قابل اجرا نیست)، مراحل کاملاً به شکل زیر انجام می‌شود:
مرحله ۱ — بارگذاری دیتاست
from sklearn.datasets import load_diabetes
data = load_diabetes()
مرحله ۲ — ساخت DataFrame
X = pd.DataFrame(data.data, columns=data.feature_names)
y = pd.Series(data.target)
مرحله ۳ — تقسیم داده‌ها

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
مرحله ۴ — ساخت مدل Linear Regression

from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)
مرحله ۵ — پیش‌بینی

y_pred = model.predict(X_test)
مرحله ۶ — محاسبه معیارهای ارزیابی

from sklearn.metrics import mean_squared_error, r2_score

mse = mean_squared_error(y_test, y_pred)
r2  = r2_score(y_test, y_pred)
۳. نتایج معمول Linear Regression روی این دیتاست

(طبق اجراهای استاندارد، معمولاً این نتایج به دست می‌آید)
• Mean Squared Error ≈ 2900
• R² Score ≈ 0.45

تفسیر:
• مقدار خطا (MSE) نسبتاً بزرگ است، چون داده‌ها پیچیده و پراکندگی زیاد دارند.
• R²=0.45 یعنی:
مدل حدود ۴۵٪ از تغییرات خطر/پیشرفت دیابت را توضیح می‌دهد
که برای داده‌های پزشکی یک مقدار قابل قبول است.
